.github/workflows/build-and-integration-tests.yml

```diff
@@ -122,30 +115,112 @@ jobs:
-      run: tar.exe -a -c -f cargo-binstall-${{ matrix.target }}.zip ${{ matrix.output }}
-
-    - name: Create archive (zip, macos)
-      if: ${{ matrix.os == 'macos-latest' }}
-      run: zip cargo-binstall-${{ matrix.target }}.zip ${{ matrix.output }}
+    - name: Upload output
+      uses: actions/upload-artifact@v3
+      with:
+        retention-days: 1
+        name: "${{ matrix.target }}.${{ matrix.output }}"
+        path: "${{ matrix.output }}"
+
+  release-test:
+    name: Package and release (testing)
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Get outputs
+      uses: actions/download-artifact@v3
+      with:
+        path: outputs/
 
-    - name: Upload artifacts
-      uses: actions/upload-artifact@v1
+    - name: Create archives
+      shell: bash
+      run: |
+        set -euxo pipefail
+        for o in outputs/*; do
+          pushd "$o"
+
+          cp ../../LICENSE.txt ../../README.md .
+
+          target=$(basename "$o" | cut -d. -f1)
+          if grep -qE '(apple|windows)' <<< "$target"; then
+            zip -9 "../cargo-binstall-${target}.zip" *
+          else
+            tar cv * | gzip -9 > "../cargo-binstall-${target}.tgz"
+          fi
+
+          popd
+        done
+    - name: Upload output
+      uses: actions/upload-artifact@v3
+      with:
+        retention-days: 1
+        name: release-bits
+        path: |
+          outputs/cargo-binstall-*.zip
+          outputs/cargo-binstall-*.tgz
+
+  release:
+    name: Package and release
+    needs: build
+    if: ${{ startsWith(github.ref, 'refs/tags/v') }}
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Get outputs
+      uses: actions/download-artifact@v3
       with:
-        name: cargo-binstall-${{ matrix.target }}.${{ matrix.archive }}
-        path: cargo-binstall-${{ matrix.target }}.${{ matrix.archive }}
+        path: outputs/
 
-    - name: Upload binary to release
-      if: ${{ startsWith(github.ref, 'refs/tags/v') }}
-      uses: svenstaro/upload-release-action@v2
+    - name: Create archives
+      shell: bash
+      run: |
+        set -euxo pipefail
+        for o in outputs/*; do
+          pushd "$o"
+
+          cp ../../LICENSE.txt ../../README.md .
+
+          target=$(basename "$o" | cut -d. -f1)
+          if grep -qE '(apple|windows)' <<< "$target"; then
+            zip "../cargo-binstall-${target}.zip" *
+          else
+            tar cvf "../cargo-binstall-${target}.tgz" *
+          fi
+
+          popd
+        done
+
+    - name: Extract release notes
+      id: notes
+      shell: bash
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        GITHUB_REPO: ${{ github.repository }}
+        release_commit: ${{ github.event.head_commit.message }}
+      run: |
+        set -euxo pipefail
+        release_pr=$(head -n1 <<< "${release_commit:-}" | grep -oP '(?<=[(]#)\d+(?=[)])')
```

Comment:
I don't like the use of regex here, it is like cipher text.

Is there any other way to extract this, like writing a python script?